File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/security/cfg/CFGBuilder.kt
/**
 * KSA-164: CFG Builder — Constructs control flow graphs from AST function nodes.
 */
package com.codeintel.analyzers.security.cfg

import com.codeintel.analyzers.security.BlockType
import com.codeintel.analyzers.security.EdgeType
import com.codeintel.parsers.SyntaxNode

class CFGBuilder {
    private var blockCounter = 0
    private lateinit var cfg: ControlFlowGraph

    fun build(functionNode: SyntaxNode, language: String): ControlFlowGraph {
        blockCounter = 0
        val entry = newBlock(BlockType.ENTRY)
        cfg = ControlFlowGraph(entry)

        val exit = newBlock(BlockType.EXIT)
        cfg.addBlock(exit)

        val body = getFunctionBody(functionNode, language)
        if (body == null) {
            cfg.addEdge(entry, exit, EdgeType.SEQUENTIAL)
            return cfg
        }

        val lastBlock = processStatements(body, entry, exit)
        if (lastBlock != null && lastBlock !== exit) {
            cfg.addEdge(lastBlock, exit, EdgeType.SEQUENTIAL)
        }
        return cfg
    }

    private fun getFunctionBody(node: SyntaxNode, language: String): SyntaxNode? {
        node.childByFieldName("body")?.let { return it }
        for (child in node.namedChildren) {
            if (child.type in listOf("statement_block", "block")) return child
        }
        return null
    }

    private fun processStatements(blockNode: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        var active: BasicBlock? = currentBlock
        for (child in blockNode.namedChildren) {
            if (active == null) break
            active = processStatement(child, active, exitBlock)
        }
        return active
    }

    private fun processStatement(stmt: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        return when (stmt.type) {
            "if_statement" -> handleIf(stmt, currentBlock, exitBlock)
            "while_statement", "do_statement" -> handleWhile(stmt, currentBlock, exitBlock)
            "for_statement", "for_in_statement" -> handleFor(stmt, currentBlock, exitBlock)
            "try_statement" -> handleTryCatch(stmt, currentBlock, exitBlock)
            "switch_statement" -> handleSwitch(stmt, currentBlock, exitBlock)
            "return_statement" -> {
                currentBlock.addStatement(stmt)
                cfg.addEdge(currentBlock, exitBlock, EdgeType.RETURN)
                null
            }
            "throw_statement" -> {
                currentBlock.addStatement(stmt)
                cfg.addEdge(currentBlock, exitBlock, EdgeType.EXCEPTION)
                null
            }
            "break_statement", "continue_statement" -> {
                currentBlock.addStatement(stmt)
                null
            }
            else -> {
                currentBlock.addStatement(stmt)
                currentBlock
            }
        }
    }

    private fun handleIf(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        node.childByFieldName("condition")?.let { currentBlock.addStatement(it) }

        val mergeBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(mergeBlock)

        val consequence = node.childByFieldName("consequence")
        val thenBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(thenBlock)
        cfg.addEdge(currentBlock, thenBlock, EdgeType.BRANCH_TRUE)

        var thenEnd: BasicBlock? = thenBlock
        if (consequence != null) thenEnd = processBlockOrStatement(consequence, thenBlock, exitBlock)
        if (thenEnd != null) cfg.addEdge(thenEnd, mergeBlock, EdgeType.SEQUENTIAL)

        val alternative = node.childByFieldName("alternative")
        if (alternative != null) {
            val elseBlock = newBlock(BlockType.NORMAL)
            cfg.addBlock(elseBlock)
            cfg.addEdge(currentBlock, elseBlock, EdgeType.BRANCH_FALSE)

            val elseBody = if (alternative.type == "else_clause" && alternative.namedChildCount > 0)
                alternative.namedChildren.first() else alternative
            var elseEnd: BasicBlock? = elseBlock
            if (elseBody.type == "if_statement") {
                elseEnd = handleIf(elseBody, elseBlock, exitBlock)
            } else {
                elseEnd = processBlockOrStatement(elseBody, elseBlock, exitBlock)
            }
            if (elseEnd != null) cfg.addEdge(elseEnd, mergeBlock, EdgeType.SEQUENTIAL)
        } else {
            cfg.addEdge(currentBlock, mergeBlock, EdgeType.BRANCH_FALSE)
        }

        return mergeBlock
    }

    private fun handleWhile(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        val headerBlock = newBlock(BlockType.LOOP_HEADER)
        cfg.addBlock(headerBlock)
        cfg.addEdge(currentBlock, headerBlock, EdgeType.SEQUENTIAL)

        node.childByFieldName("condition")?.let { headerBlock.addStatement(it) }

        val postLoop = newBlock(BlockType.NORMAL)
        cfg.addBlock(postLoop)

        val body = node.childByFieldName("body")
        val bodyBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(bodyBlock)
        cfg.addEdge(headerBlock, bodyBlock, EdgeType.BRANCH_TRUE)
        cfg.addEdge(headerBlock, postLoop, EdgeType.LOOP_EXIT)

        var bodyEnd: BasicBlock? = bodyBlock
        if (body != null) bodyEnd = processBlockOrStatement(body, bodyBlock, exitBlock)
        if (bodyEnd != null) cfg.addEdge(bodyEnd, headerBlock, EdgeType.LOOP_BACK)

        return postLoop
    }

    private fun handleFor(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        node.childByFieldName("initializer")?.let { currentBlock.addStatement(it) }

        val headerBlock = newBlock(BlockType.LOOP_HEADER)
        cfg.addBlock(headerBlock)
        cfg.addEdge(currentBlock, headerBlock, EdgeType.SEQUENTIAL)

        node.childByFieldName("condition")?.let { headerBlock.addStatement(it) }

        val postLoop = newBlock(BlockType.NORMAL)
        cfg.addBlock(postLoop)

        val body = node.childByFieldName("body")
        val bodyBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(bodyBlock)
        cfg.addEdge(headerBlock, bodyBlock, EdgeType.BRANCH_TRUE)
        cfg.addEdge(headerBlock, postLoop, EdgeType.LOOP_EXIT)

        var bodyEnd: BasicBlock? = bodyBlock
        if (body != null) bodyEnd = processBlockOrStatement(body, bodyBlock, exitBlock)

        node.childByFieldName("increment")?.let { if (bodyEnd != null) bodyEnd!!.addStatement(it) }
        if (bodyEnd != null) cfg.addEdge(bodyEnd!!, headerBlock, EdgeType.LOOP_BACK)

        return postLoop
    }

    private fun handleTryCatch(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        val mergeBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(mergeBlock)

        val tryBody = node.childByFieldName("body")
        val tryBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(tryBlock)
        cfg.addEdge(currentBlock, tryBlock, EdgeType.SEQUENTIAL)

        var tryEnd: BasicBlock? = tryBlock
        if (tryBody != null) tryEnd = processBlockOrStatement(tryBody, tryBlock, exitBlock)
        if (tryEnd != null) cfg.addEdge(tryEnd, mergeBlock, EdgeType.SEQUENTIAL)

        val handler = node.childByFieldName("handler")
        if (handler != null) {
            val catchBlock = newBlock(BlockType.CATCH)
            cfg.addBlock(catchBlock)
            cfg.addEdge(tryBlock, catchBlock, EdgeType.EXCEPTION)

            val catchBody = handler.childByFieldName("body")
            var catchEnd: BasicBlock? = catchBlock
            if (catchBody != null) catchEnd = processBlockOrStatement(catchBody, catchBlock, exitBlock)
            if (catchEnd != null) cfg.addEdge(catchEnd, mergeBlock, EdgeType.SEQUENTIAL)
        }

        val finalizer = node.childByFieldName("finalizer")
        if (finalizer != null) {
            val finallyBlock = newBlock(BlockType.NORMAL)
            cfg.addBlock(finallyBlock)
            cfg.addEdge(mergeBlock, finallyBlock, EdgeType.SEQUENTIAL)
            if (finalizer.namedChildCount > 0) {
                processBlockOrStatement(finalizer.namedChildren.first(), finallyBlock, exitBlock)
            }
            return finallyBlock
        }

        return mergeBlock
    }

    private fun handleSwitch(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        node.childByFieldName("value")?.let { currentBlock.addStatement(it) }

        val mergeBlock = newBlock(BlockType.NORMAL)
        cfg.addBlock(mergeBlock)

        val body = node.childByFieldName("body") ?: return mergeBlock

        for (caseNode in body.namedChildren) {
            val caseBlock = newBlock(BlockType.NORMAL)
            cfg.addBlock(caseBlock)
            cfg.addEdge(currentBlock, caseBlock, EdgeType.BRANCH_TRUE)

            var caseEnd: BasicBlock? = caseBlock
            for (child in caseNode.namedChildren) {
                if (caseEnd == null) break
                if (child.type in listOf("switch_case", "switch_default")) continue
                caseEnd = processStatement(child, caseEnd, exitBlock)
            }
            if (caseEnd != null) cfg.addEdge(caseEnd, mergeBlock, EdgeType.SEQUENTIAL)
        }

        return mergeBlock
    }

    private fun processBlockOrStatement(node: SyntaxNode, currentBlock: BasicBlock, exitBlock: BasicBlock): BasicBlock? {
        return if (node.type in listOf("statement_block", "block")) {
            processStatements(node, currentBlock, exitBlock)
        } else {
            processStatement(node, currentBlock, exitBlock)
        }
    }

    private fun newBlock(type: BlockType): BasicBlock {
        return BasicBlock(blockCounter++, type)
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/security/cfg/cfg_builder.py
"""
KSA-164: CFG Builder — Constructs control flow graphs from AST function nodes.
Handles if/else, loops, try/catch, switch, and early returns.
"""

from __future__ import annotations
from typing import Any, Optional
from .basic_block import BasicBlock
from .control_flow_graph import ControlFlowGraph


class CFGBuilder:
    def __init__(self) -> None:
        self._block_counter = 0
        self._cfg: Optional[ControlFlowGraph] = None

    def build(self, function_node: Any, language: str) -> ControlFlowGraph:
        """Build CFG from a function AST node."""
        self._block_counter = 0

        entry = self._new_block("entry")
        self._cfg = ControlFlowGraph(entry)

        exit_block = self._new_block("exit")
        self._cfg.add_block(exit_block)

        body = self._get_function_body(function_node, language)
        if not body:
            self._cfg.add_edge(entry, exit_block, "sequential")
            return self._cfg

        last_block = self._process_statements(body, entry, exit_block)
        if last_block and last_block is not exit_block:
            self._cfg.add_edge(last_block, exit_block, "sequential")

        return self._cfg

    def _get_function_body(self, node: Any, language: str) -> Optional[Any]:
        body = node.child_by_field_name("body")
        if body:
            return body
        for child in node.named_children:
            if child.type in ("statement_block", "block"):
                return child
        return None

    def _process_statements(self, block_node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        active: Optional[BasicBlock] = current_block
        for child in block_node.named_children:
            if not active:
                break
            active = self._process_statement(child, active, exit_block)
        return active

    def _process_statement(self, stmt: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        stype = stmt.type
        if stype == "if_statement":
            return self._handle_if(stmt, current_block, exit_block)
        elif stype in ("while_statement", "do_statement"):
            return self._handle_while(stmt, current_block, exit_block)
        elif stype in ("for_statement", "for_in_statement"):
            return self._handle_for(stmt, current_block, exit_block)
        elif stype == "try_statement":
            return self._handle_try_catch(stmt, current_block, exit_block)
        elif stype == "switch_statement":
            return self._handle_switch(stmt, current_block, exit_block)
        elif stype == "return_statement":
            current_block.add_statement(stmt)
            self._cfg.add_edge(current_block, exit_block, "return")
            return None
        elif stype == "throw_statement":
            current_block.add_statement(stmt)
            self._cfg.add_edge(current_block, exit_block, "exception")
            return None
        elif stype in ("break_statement", "continue_statement"):
            current_block.add_statement(stmt)
            return None
        else:
            current_block.add_statement(stmt)
            return current_block

    def _handle_if(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        cond_node = node.child_by_field_name("condition")
        if cond_node:
            current_block.add_statement(cond_node)

        merge_block = self._new_block("normal")
        self._cfg.add_block(merge_block)

        consequence = node.child_by_field_name("consequence")
        then_block = self._new_block("normal")
        self._cfg.add_block(then_block)
        self._cfg.add_edge(current_block, then_block, "branch-true")

        then_end: Optional[BasicBlock] = then_block
        if consequence:
            then_end = self._process_block_or_statement(consequence, then_block, exit_block)
        if then_end:
            self._cfg.add_edge(then_end, merge_block, "sequential")

        alternative = node.child_by_field_name("alternative")
        if alternative:
            else_block = self._new_block("normal")
            self._cfg.add_block(else_block)
            self._cfg.add_edge(current_block, else_block, "branch-false")

            else_body = alternative.named_children[0] if alternative.type == "else_clause" and alternative.named_child_count > 0 else alternative
            else_end: Optional[BasicBlock] = else_block
            if else_body:
                if else_body.type == "if_statement":
                    else_end = self._handle_if(else_body, else_block, exit_block)
                else:
                    else_end = self._process_block_or_statement(else_body, else_block, exit_block)
            if else_end:
                self._cfg.add_edge(else_end, merge_block, "sequential")
        else:
            self._cfg.add_edge(current_block, merge_block, "branch-false")

        return merge_block

    def _handle_while(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        header_block = self._new_block("loop-header")
        self._cfg.add_block(header_block)
        self._cfg.add_edge(current_block, header_block, "sequential")

        cond_node = node.child_by_field_name("condition")
        if cond_node:
            header_block.add_statement(cond_node)

        post_loop = self._new_block("normal")
        self._cfg.add_block(post_loop)

        body = node.child_by_field_name("body")
        body_block = self._new_block("normal")
        self._cfg.add_block(body_block)
        self._cfg.add_edge(header_block, body_block, "branch-true")
        self._cfg.add_edge(header_block, post_loop, "loop-exit")

        body_end: Optional[BasicBlock] = body_block
        if body:
            body_end = self._process_block_or_statement(body, body_block, exit_block)
        if body_end:
            self._cfg.add_edge(body_end, header_block, "loop-back")

        return post_loop

    def _handle_for(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        init = node.child_by_field_name("initializer")
        if init:
            current_block.add_statement(init)

        header_block = self._new_block("loop-header")
        self._cfg.add_block(header_block)
        self._cfg.add_edge(current_block, header_block, "sequential")

        cond_node = node.child_by_field_name("condition")
        if cond_node:
            header_block.add_statement(cond_node)

        post_loop = self._new_block("normal")
        self._cfg.add_block(post_loop)

        body = node.child_by_field_name("body")
        body_block = self._new_block("normal")
        self._cfg.add_block(body_block)
        self._cfg.add_edge(header_block, body_block, "branch-true")
        self._cfg.add_edge(header_block, post_loop, "loop-exit")

        body_end: Optional[BasicBlock] = body_block
        if body:
            body_end = self._process_block_or_statement(body, body_block, exit_block)

        increment = node.child_by_field_name("increment")
        if increment and body_end:
            body_end.add_statement(increment)
        if body_end:
            self._cfg.add_edge(body_end, header_block, "loop-back")

        return post_loop

    def _handle_try_catch(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        merge_block = self._new_block("normal")
        self._cfg.add_block(merge_block)

        try_body = node.child_by_field_name("body")
        try_block = self._new_block("normal")
        self._cfg.add_block(try_block)
        self._cfg.add_edge(current_block, try_block, "sequential")

        try_end: Optional[BasicBlock] = try_block
        if try_body:
            try_end = self._process_block_or_statement(try_body, try_block, exit_block)
        if try_end:
            self._cfg.add_edge(try_end, merge_block, "sequential")

        handler = node.child_by_field_name("handler")
        if handler:
            catch_block = self._new_block("catch")
            self._cfg.add_block(catch_block)
            self._cfg.add_edge(try_block, catch_block, "exception")

            catch_body = handler.child_by_field_name("body")
            catch_end: Optional[BasicBlock] = catch_block
            if catch_body:
                catch_end = self._process_block_or_statement(catch_body, catch_block, exit_block)
            if catch_end:
                self._cfg.add_edge(catch_end, merge_block, "sequential")

        finalizer = node.child_by_field_name("finalizer")
        if finalizer:
            finally_block = self._new_block("normal")
            self._cfg.add_block(finally_block)
            self._cfg.add_edge(merge_block, finally_block, "sequential")
            finally_body = finalizer.named_children[0] if finalizer.named_child_count > 0 else None
            if finally_body:
                self._process_block_or_statement(finally_body, finally_block, exit_block)
            return finally_block

        return merge_block

    def _handle_switch(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        value = node.child_by_field_name("value")
        if value:
            current_block.add_statement(value)

        merge_block = self._new_block("normal")
        self._cfg.add_block(merge_block)

        body = node.child_by_field_name("body")
        if not body:
            return merge_block

        for case_node in body.named_children:
            case_block = self._new_block("normal")
            self._cfg.add_block(case_block)
            self._cfg.add_edge(current_block, case_block, "branch-true")

            case_end: Optional[BasicBlock] = case_block
            for child in case_node.named_children:
                if not case_end:
                    break
                if child.type in ("switch_case", "switch_default"):
                    continue
                case_end = self._process_statement(child, case_end, exit_block)
            if case_end:
                self._cfg.add_edge(case_end, merge_block, "sequential")

        return merge_block

    def _process_block_or_statement(self, node: Any, current_block: BasicBlock, exit_block: BasicBlock) -> Optional[BasicBlock]:
        if node.type in ("statement_block", "block"):
            return self._process_statements(node, current_block, exit_block)
        return self._process_statement(node, current_block, exit_block)

    def _new_block(self, block_type: str) -> BasicBlock:
        block = BasicBlock(self._block_counter, block_type)
        self._block_counter += 1
        return block
